### PR TITLE
Muzzle flash

### PR DESCRIPTION
**Pull request by Night-Pryanik**
#### Summary
Features "Muzzle flash"

#### Purpose of change
Firearms IRL generate muzzle flash when firing.

#### Describe the solution
- Added new `fd_muzzle_flash` field with `half_life` of 1 second.
- This field will be generated on position of the gun, unless said gun have a flash hider or suppressor installed on it.
- Most suppressors got the `FLASH_HIDER` flag and will eliminate muzzle flash from a shot.
- The field will be removed on the next turn.

#### Describe alternatives you've considered
None.

#### Testing
Made sure the time is 22:00. Got myself M4A1. Fired at debug monster - semi-auto and full-auto. Observed muzzle flash which illuminates the position of the shooter and lasts for one turn.
Installed suppressor on the gun. Fired at debug monster. No muzzle flash was generated.
Fired an arrow from a composite bow. No muzzle flash was generated.

#### Additional context

There is some issue with field sometimes lasting for two turns instead of one - for example, it's seen on last shot in gif. I suppose that's something related to move cost of firing which doesn't remove the field in time. I tried my best, but unfortunately I couldn't eliminate this strange "delay". I'm not sure if this is a big deal and if it's tolerable enough. If it's not, I'd appreciate any help on this matter.

### CODE REVIEW

File: src/ranged.cpp [hunk 1]
```diff
@@ -99,6 +99,8 @@ static const fault_id fault_gun_blackpowder( "fault_gun_blackpowder" );
 static const fault_id fault_gun_chamber_spent( "fault_gun_chamber_spent" );
 static const fault_id fault_gun_dirt( "fault_gun_dirt" );
 
+const flag_id flag_FLASH_HIDER( "FLASH_HIDER" );
```
Comment (BrettDong) [inline src/ranged.cpp:102]:
```suggestion
static const flag_id json_flag_FLASH_HIDER( "FLASH_HIDER" );
```

**Comment by anothersimulacrum** [comment]
Going off that gif there, it seems quite bright to illuminate a 5x5 area from a muzzle flash. I'd expect something more like 1x1.

**Comment by Night-Pryanik** [comment]
Ok, lowered radius to 3x3. 1x1 means it's only illuminating the gunner's tile, and it wasn't looking good.

**Comment by souricelle** [comment]
A 1x1 illumination is the same as a laptop screen, phone flashlight, or candle, which [seems completely appropriate to me.](https://www.youtube.com/watch?v=2GN17VCS2gs)

**Comment by Night-Pryanik** [comment]
Your video example shows the flash from a handgun. Flash from a rifle is way more brighter, see [this video for example](https://www.youtube.com/watch?v=P1xpqL04cA8).

Screenshot from the aforementioned video
изображение

**Comment by anothersimulacrum** [comment]
Perhaps the field should have different intensity levels based on the ammunition?

**Comment by Night-Pryanik** [comment]
> Perhaps the field should have different intensity levels based on the ammunition?

Yeah, I thought about it too, but I don't want to make this PR too complex. It could be done in subsequent PRs.

**Comment by souricelle** [comment]
Then probably it should err on the dimmer side yeah?

File: src/field.cpp [hunk 2]
```diff
@@ -86,6 +86,13 @@ void field_entry::do_decay()
             }
             return;
         }
+
+        // Special case for fields that need to exist no more than one turn and thus have their half_life set to minimal value of 1 second, like muzzle flash:
+        if( type.obj().half_life == 1_seconds ) {
```
Comment (anothersimulacrum) [inline src/field.cpp:91]:
```suggestion
        if( type->half_life == 1_seconds ) {
```

**Comment by gkarfakis19** [comment]
Does this mean that characters with nightvision will be constantly blinded by their own muzzle flashes when firing in the dark?
Realistic but quite unfun in my opinion.

**Comment by Night-Pryanik** [comment]
> Does this mean that characters with nightvision will be constantly blinded by their own muzzle flashes when firing in the dark?

I haven't tested this case, but I think even if such characters will be blinded by the flash, their blindness should fade on the next turn. In any case, equipping suppressor should resolve this issue.

> Dealing with hordes is going to be a literal pain as you will have to pretty much wait at least a turn between every round you fire.

Even more pain, as monsters need only one turn to notice characters illuminated by muzzle flash, and even if flash fade on the next turn, monsters could start moving to the last seen location of the gunner, even if they don't see him anymore.

**Comment by gkarfakis19** [comment]
> 
> 
> > Does this mean that characters with nightvision will be constantly blinded by their own muzzle flashes when firing in the dark?
> 
> I haven't tested this case, but I think even if such characters will be blinded by the flash, their blindness should fade on the next turn. In any case, equipping suppressor should resolve this issue.
> 
> > Dealing with hordes is going to be a literal pain as you will have to pretty much wait at least a turn between every round you fire.
> 
> Even more pain, as monsters need only one turn to notice characters illuminated by muzzle flash, and even if flash fade on the next turn, monsters could start moving to the last seen location of the gunner, even if they don't see him anymore.

Although I do not mind the game being harder because of this change, I am mostly speaking in terms of the actual user experience. Having to press TAB after every shot will get tiring very quickly.

**Comment by Night-Pryanik** [comment]
> Having to press TAB after every shot will get tiring very quickly.

As I said, even if your character is illuminated for just a single turn, it's enough for monsters to remember your position and react to that. You don't need to wait for flash to fade as monsters already have seen you and are coming for you, even if you stop firing after just one shot. So you either continue shooting as if there's no flash at all, or you're rapidly changing positions and not doing this "TABbing".

Either way, you want to shoot safely at night? Then you should find and install a suppressor.

If this still doesn't convince you, what are you suggesting?

**Comment by gkarfakis19** [comment]
> 
> 
> > Having to press TAB after every shot will get tiring very quickly.
> 
> As I said, even if your character is illuminated for just a single turn, it's enough for monsters to remember your position and react to that. You don't need to wait for flash to fade as monsters already have seen you and are coming for you, even if you stop firing after just one shot. So you either continue shooting as if there's no flash at all, or you're rapidly changing positions and not doing this "TABbing".
> 
> Either way, you want to shoot safely at night? Then you should find and install a suppressor.
> 
> If this still doesn't convince you, what are you suggesting?

I am not suggesting any changes, because I truly cannot see how this can be fixed short of scrapping the whole idea. 
In regards to my previous point, I was not speaking about monsters spotting the PC. In fact, shooting at night without a suppressor means you will be spotted by sound alone, anyway. I was talking about being constantly blinded by the shots and being unable to shoot any zombies for a turn after you fire a gun. You will then have to TAB once (or twice, assuming that listed bug doesn't get fixed) so that you can spot enemies again, which is going to be unfun.

**Comment by Night-Pryanik** [comment]
> I was talking about being constantly blinded by the shots

This might be a rather serious issue. I've never had a chance to shoot at night, so I can't tell if shooter is being blinded by its own shots. Are here anyone with such experience?

**Comment by actual-nh** [comment]
> > I was talking about being constantly blinded by the shots
> 
> This might be a rather serious issue. I've never had a chance to shoot at night, so I can't tell if shooter is being blinded by its own shots. Are here anyone with such experience?

In-game or RL?

**Comment by Night-Pryanik** [comment]
> In-game or RL?

IRL.

**Comment by actual-nh** [comment]
<https://arxiv.org/pdf/1410.6025.pdf> may be of use. It notes military weapons using additives to the powder to suppress night flash (as well as using flash suppressors; I seem to recall one such suppressor that also acted as a rifle grenade launching "cup").

**Comment by souricelle** [comment]
> > > Having to press TAB after every shot will get tiring very quickly.
> > 
> > 
> > As I said, even if your character is illuminated for just a single turn, it's enough for monsters to remember your position and react to that. You don't need to wait for flash to fade as monsters already have seen you and are coming for you, even if you stop firing after just one shot. So you either continue shooting as if there's no flash at all, or you're rapidly changing positions and not doing this "TABbing".
> > Either way, you want to shoot safely at night? Then you should find and install a suppressor.
> > If this still doesn't convince you, what are you suggesting?
> 
> I am not suggesting any changes, because I truly cannot see how this can be fixed short of scrapping the whole idea.
> In regards to my previous point, I was not speaking about monsters spotting the PC. In fact, shooting at night without a suppressor means you will be spotted by sound alone, anyway. I was talking about being constantly blinded by the shots and being unable to shoot any zombies for a turn after you fire a gun. You will then have to TAB once (or twice, assuming that listed bug doesn't get fixed) so that you can spot enemies again, which is going to be unfun.

This is a pretty big overreaction. If you want to take advantage of stealth by sticking to the dark, shooting an unsuppressed gun probably isn't what you should be doing. Get a bow, a suppressor, some throwing knives, or go melee. Having to pass a single turn so your darkvision works again is realistic and not annoying.

**Comment by ANickelN** [comment]
> > I was talking about being constantly blinded by the shots
> 
> This might be a rather serious issue. I've never had a chance to shoot at night, so I can't tell if shooter is being blinded by its own shots. Are here anyone with such experience?

I shoot at night every so often, never had issues with being blinded by my own muzzle-flash on handguns or rifles.  If I was dumping an entire magazine in full-auto fire the strobe might get me a bit but you're usually focused on the target and the light won't mess with you.

**Comment by souricelle** [comment]
You're not actually being blinded in the game in this case though, you just don't quite have your full darkvision in the split second the flash goes off.

**Comment by kevingranade** [comment]
I think the vision system isn't working as intended here.  I was testing vs a zombie edited to have no hearing or scent, and as written, this would only cause the zombie to detect me from another few tiles away, which isn't really the impact this should have, but I think making it increase the detection range by any more than that would require a prohibitively high degree of brightness for the flash.  Not sure what needs to happen, but as written this has almost no impact on monster ability to detect you, and a major impact on your own vision if you have any kind of night vision, which is not a great outcome.

**Comment by Night-Pryanik** [comment]
I'm not sure what do you mean by "few tiles", but in my tests (versus ordinary zombies, but my character has debug no scent mutation) monsters detected me at least 10 tiles away. Maybe it was indeed the sound of a fire though.
Temporary blindness is a valid gameplay issue, I'm even on the edge of closing the PR because of it.

**Comment by kevingranade** [comment]
When I tested with the PR as is, it only extended detection range for a zombie from about 5 tiles to 9 tiles, that's not really what I would expect, especially with the light as bright as it is.  What I expect to happen is drawing attention from zombies at the limit of their detection range, which should be 40 based on the default of vision_day in mtype.h

What I'm thinking could work is having the muzzle flash field set to a very low value to prevent the night blindness effect, and modify how it works so that you get illumination of adjacent tiles so you get the flash effect you want, and address whatever is causing detection range to be unexpectedly short.

**Pull request closed without merging**